### ISSUE TITLE
Plugins Diskspace fix in the Jenkins Image

### ISSUE THREAD

**Issue by waveywaves** [openshift/jenkins#802]
The Jenkins Image upon running is using a lot more disk space as the plugins are being copied directly. We have reduced the size of the builder image by using symlink of the plugins folder instead of copying the plugins. These changes can be seen in the PR below.

[fabric8-jenkins/jenkins-openshift-base#52](https://github.com/fabric8-jenkins/jenkins-openshift-base/pull/52)

We want to add this patch to upstream as this may help other developers also. How can we get this added? Should we add an option or a flag to allow developers to choose if they want to use the symlink or copy the plugins.

cc @hrishin @sthaha .

**Comment by gabemontero** [comment]
Hi @waveywaves @hrishin @sthaha 

Thins are bit more complicated for our base image than what is done with https://github.com/fabric8-jenkins/jenkins-openshift-base/pull/52 because of 

a) our current support of both RHEL and CENTOS
b) allowing users to extend the image via multiple means
c) preserving prior decisions when PVs are in play

Please refresh my memory wrt the fabric8-jenkins image and a) and b).

We actually leverage `ln` as much as possible in  https://github.com/openshift/jenkins/blob/master/2/contrib/s2i/run today

Also, things are evolving with openshift 4.0 and a) that *might* lead to some simplifying assumptions down the line (post 4.0).

Bottom line this is most likely complicated enough to warrant some form of design/exchange/review.

@bparees @adambkaplan @pweil-

**Comment by grdryn** [comment]
I just came here to add a link to #735 as an additional thing to think about around how plugins are handled in this image. :man_cartwheeling:

**Comment by hrishin** [comment]
Hi @gabemontero 

> a) our current support of both RHEL and CENTOS

Intent means, maintain the separate patch for both images. Please, could you elaborate more?

> b) allowing users to extend the image via multiple means

Yes, with such kind of symlinking users are still extent it there way. If they override `run` script complete its still their choice.
However, we would like to enable this symlinking only if certain ENV var do exist with a certain value. Otherwise, the image would behave as current behavior. So the user can still extent behavior by using ENV var as a flag.

> c) preserving prior decisions when PVs are in play

I certainly agree with this. However, this kind of optimization is really needed for OSIO kind of use case. 

Can see `ln` used in the upstream here https://github.com/openshift/jenkins/blob/master/2/contrib/s2i/run#L306 . Thank you for pointing out, we tried a [similar approach](https://github.com/fabric8-jenkins/jenkins-openshift-base/pull/49/files) as well. However, https://github.com/fabric8-jenkins/jenkins-openshift-base/pull/52/files turned out better in our observation. 

> Bottom line this is most likely complicated enough to warrant some form of design/exchange/review.

Sure, we can disucss it in detail.

**Comment by gabemontero** [comment]
OK @hrishin  ... elaboration:
- As you all are probably aware, with each new release of openshift, we cut a new tag for the openshift jenkins images.  So we have a v3.6 tag, v3.7 tag (3.8 was never made fully supported, it was skipped), v3.9, v3.10, and v3.11
- v3.11 is the last v3.x release; we are in the midst of 4.0 development.
- We provide 2 types each openshift jenkins image
1) the CentOS7 version;  this is our "community" or "OKD" based version;  it serves as our on-going pre-release version for a given OpenShift release cycle;  So there are v4.0 tagged versions of these images available;  docker.io was traditionally the repository for those, but with 4.0, we are switching to quay.io;  The CentOS image is only supported in a best effort, community fashion, sort of way (github issues, etc.)
2) the RHEL7 version;  this is what paying customers with RH subscriptions get;  The "OCP" or fully supported version of openshift.  Customers with problems open support cases, bugzillas (in our case) get open, and they get full support.  We only push new tags when an openshift release ships.  Then, after a release, when errata or zfixes occur, new versions of say the v.311 tag show up.  Those images are only available at the Red Hat registries.  Used to be registry.access.redhat.com (which is still freely avaiable), but we are moving to the terms based registry registry.redhat.io, where you have to pay to get a token to be able to log into the registry and access the content

- RH has a lot of rules (usually security based in scope) for the content of RHEL images.  Content can only be installed via RPMs and yum/dnf.  And in building the images, the build systems cannot access the internet.  So content like plugins has to be provided in dist git before the build system actually produces the image.

- the fly in the ointment is that jenkins allows you to update plugins via their UI and the jenkins update center;  and of course we allow you to extend our image.  But we needed a means that support could consume to determine when a customer overrides the set of plugins we provided (and what we scope support to) when a version of openshift GAs.

We use the fact whether a plugin file in the jenkins dir is a link or an actual file for that.  If they don't override anything, support can log into a pod running jenkins and see the plugin files are links to the RPM content.  If they replaced the plugin, then the file is no longer a link.

- all this is further muddied by the fact that in 4.0, there is shifting away from CentOS based images.

- So we need to reconcile these above particulars with the particulars of what images you produce /support .... that reconciliation is non-trivial, and beyond the scope of the fabric PR, IMO

Also beyond @grdryn 's note at first blush I think.

@bparees @adambkaplan @pweil-

**Comment by gabemontero** [comment]
Minor correction from @bparees:

you just have to agree to the terms of use to get a token to access registry.redhat.io images
but of course if you want support for using those images you need to pay

**Comment by hrishin** [comment]
Thank you @gabemontero for the detailed background.

**Comment by waveywaves** [comment]
Closing this issue for now as it isn't urgent as of now and a lot of things need to factored in based on https://github.com/openshift/jenkins/issues/802#issuecomment-466042475
